flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java

```diff
@@ -1662,6 +1665,82 @@ public void testTerminationOnFatalError() throws Throwable {
+
+		final CompletableFuture<AllocationID> allocationNotifiedFree = new CompletableFuture<>();
+
+		final CompletableFuture<Void> initialSlotReportFuture = new CompletableFuture<>();
+		testingResourceManagerGateway.setSendSlotReportFunction(resourceIDInstanceIDSlotReportTuple3 -> {
+			initialSlotReportFuture.complete(null);
+			return CompletableFuture.completedFuture(Acknowledge.get());
+
+		});
+
+		testingResourceManagerGateway.setNotifySlotAvailableConsumer(instanceIDSlotIDAllocationIDTuple3 ->
+			allocationNotifiedFree.complete(instanceIDSlotIDAllocationIDTuple3.f2));
+
+		rpc.registerGateway(testingResourceManagerGateway.getAddress(), testingResourceManagerGateway);
+		resourceManagerLeaderRetriever.notifyListener(testingResourceManagerGateway.getAddress(), testingResourceManagerGateway.getFencingToken().toUUID());
+
+		final CountDownLatch slotOfferings = new CountDownLatch(2);
+		final CompletableFuture<AllocationID> failedSlotFuture = new CompletableFuture<>();
+		final TestingJobMasterGateway jobMasterGateway = new TestingJobMasterGatewayBuilder()
+				.setFailSlotConsumer((resourceID, allocationID, throwable) -> {
+					failedSlotFuture.complete(allocationID);
+				})
+				.setOfferSlotsFunction((resourceID, slotOffers) -> {
+					slotOffers.forEach((ignored) -> slotOfferings.countDown());
+					return CompletableFuture.completedFuture(new ArrayList<>(slotOffers));
+				})
+				.build();
+		final String jobManagerAddress = jobMasterGateway.getAddress();
+		rpc.registerGateway(jobManagerAddress, jobMasterGateway);
+		jobManagerLeaderRetriever.notifyListener(jobManagerAddress, jobMasterGateway.getFencingToken().toUUID());
+
+		taskExecutor.start();
+
+		try {
+			final TaskExecutorGateway taskExecutorGateway = taskExecutor.getSelfGateway(TaskExecutorGateway.class);
+
+			initialSlotReportFuture.get();
+
+			final SlotID slotId1 = new SlotID(taskExecutor.getResourceID(), 0);
+			final SlotID slotId2 = new SlotID(taskExecutor.getResourceID(), 1);
+			final AllocationID allocationIdInBoth = new AllocationID();
+			final AllocationID allocationIdOnlyInJM = new AllocationID();
+			final AllocationID allocationIdOnlyInTM = new AllocationID();
+
+			taskExecutorGateway.requestSlot(slotId1, jobId, allocationIdInBoth, "foobar", testingResourceManagerGateway.getFencingToken(), timeout);
+			taskExecutorGateway.requestSlot(slotId2, jobId, allocationIdOnlyInTM, "foobar", testingResourceManagerGateway.getFencingToken(), timeout);
+
+			slotOfferings.await();
+
+			List<OfferedSlot> offeredSlots = Arrays.asList(
+					new OfferedSlot(0, allocationIdInBoth),
+					new OfferedSlot(1, allocationIdOnlyInJM)
+			);
+			SlotSnapshot snapshot = new SlotSnapshot(jobId, offeredSlots);
+			taskExecutorGateway.heartbeatFromJobManager(ResourceID.generate(), snapshot);
+
+			assertEquals(allocationIdOnlyInJM, failedSlotFuture.get());
+			assertEquals(allocationIdOnlyInTM, allocationNotifiedFree.get());
```

Comment:
Could use `assertThat` to use Hamcrest matchers here.

Reply:
OK